### PR TITLE
Search preview panel: use static highlight ranges from search results

### PR DESCRIPTION
**Pull request by camdencheek**
Not all searches can be represented as a single regex pattern, so we can't easily translate a search query into a code mirror search that shows the same results. However, we already have the set of matched ranges from our search results, so we can use that to provide a static set of highlights in the preview panel.

This PR does exactly that. Note that it implements the change within the CodeMirror extensions framework, so it won't require much change to get it working in svelte as well.

[Demo loom](https://www.loom.com/share/f53b9dab21404d9a8d06e173addc1f7e)

~Remaining TODOs:~
1) ~We should scroll the preview panel to the first match. I couldn't quite figure out how to apply an effect on extension application though, so I could use some help there.~
2) ~A nice-to-have would be to retain the search panel's previous/next functionality even if the ranges are static~

## Test plan

Manual testing. What would be the appropriate types of tests to add for a PR like this?

### CODE REVIEW

File: client/web/src/repo/blob/codemirror/static-highlights.tsx [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+function findSelectedIdx(ranges: HighlightedRange[]): number | undefined {
+    const selectedIdx = ranges.findIndex(range => range.selected)
+    return selectedIdx === -1 ? undefined : selectedIdx
+}
+
+export const STATIC_HIGHLIGHTS_CONTAINER_ID = 'static-highlights-navigation-container'
+
+class StaticHighlightsPanel implements Panel {
```
Comment (camdencheek) [inline client/web/src/repo/blob/codemirror/static-highlights.tsx:139]: The panel structure is largely copy/pasted from the search panel. I just heavily pruned it to only the relevant pieces.

File: client/web/src/repo/blob/codemirror/static-highlights.tsx [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+                    from: toCodeMirrorLocation(state, range.start),
+                    to: toCodeMirrorLocation(state, range.end),
+                }))
+            ),
+            showPanel.of(view => new StaticHighlightsPanel(view, navigate)),
+            scrollToFirstRange(),
+        ],
+    })
+    return facet.of(ranges)
+}
```
Comment (camdencheek) [inline client/web/src/repo/blob/codemirror/static-highlights.tsx:54]: This is the primary entrypoint: a facet that provides four extensions: a theme, state, a panel, and a helper that scrolls to the first match on open.

File: client/web/src/repo/blob/codemirror/static-highlights.tsx [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+/**
+ * scrollToFirstRange is a small utility extension that just scrolls to
+ * the selected range on startup.
+ */
+function scrollToFirstRange(): Extension {
```
Comment (camdencheek) [inline client/web/src/repo/blob/codemirror/static-highlights.tsx:60]: This felt a little hacky, but it was the best way I could figure out how to do it. If anyone knows of a better way, let me know.

File: client/web/src/repo/blob/codemirror/static-highlights.tsx [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+            Decoration.set(
+                ranges.map(({ selected, from, to }) =>
+                    selected
+                        ? staticHighlightSelectedDecoration.range(from, to)
+                        : staticHighlightDecoration.range(from, to)
+                ),
+                true
+            )
+        ),
+})
```
Comment (camdencheek) [inline client/web/src/repo/blob/codemirror/static-highlights.tsx:116]: The state calculates the current selected range from the main selection, then uses that information to provide a set of decorations.

File: client/web/src/repo/blob/codemirror/static-highlights.tsx [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+const staticHighlightTheme = EditorView.theme({
+    '.cm-sg-static-highlight': {
+        backgroundColor: 'var(--mark-bg)',
+    },
+    '.cm-sg-static-highlight-selected': {
+        backgroundColor: 'var(--oc-orange-3)',
+    },
```
Comment (camdencheek) [inline client/web/src/repo/blob/codemirror/static-highlights.tsx:124]: Copied from the search implementation.

File: client/web/src/repo/blob/codemirror/static-highlights.tsx (hunk 3 above)
Comment (vovakulikov) [reply client/web/src/repo/blob/codemirror/static-highlights.tsx:60]: If I understood this correctly the problem comes from the fact that we don't have a good hook/place to run/dispatch scrollToAction for the panel. This is something that I also struggled with when I was trying 
to do a first implementation of the panel. If you run this too soon in constructor it won't work obviously but if it's too late you will see jump state from line 0 to line what you need to scroll 

cc @fkling maybe you know the idiomatic code mirror way to scroll to the line logic?

File: client/web/src/repo/blob/codemirror/static-highlights.tsx [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+ * staticHighlights is an extension that highlights a static set of ranges
+ * and opens a panel that allows navigation between these highlights.
+ */
+export function staticHighlights(navigate: NavigateFunction, ranges: Range[]): Extension {
+    const facet = Facet.define<Range[], Range[]>({
+        combine: ranges => ranges.flat(),
+        enables: () => [
+            staticHighlightTheme,
+            staticHighlightState.init(state =>
+                sortBy(ranges, ['from']).map((range, i) => ({
```
Comment (vovakulikov) [inline client/web/src/repo/blob/codemirror/static-highlights.tsx:43]: Not sure this is a problem here or something else, but sorting probably doesn't work as expected in some cases, for example here the 

https://github.com/sourcegraph/sourcegraph/assets/18492575/1c95e8a9-f49b-4cdc-886f-38d6dc71183a

**Review by vovakulikov — COMMENTED** [review]
Found one problem with matches ordering

File: client/web/src/repo/blob/codemirror/static-highlights.tsx (hunk 6 above)
Comment (camdencheek) [reply client/web/src/repo/blob/codemirror/static-highlights.tsx:43]: Yep, that's exactly what it is. I'll just sort after the map.

Comment (camdencheek) [reply client/web/src/repo/blob/codemirror/static-highlights.tsx]: Fixed in the latest commit 👍

**Review by vovakulikov — APPROVED** [review]

**Pull request merged**